### PR TITLE
Prevalidate productlist before attaching to api

### PR DESCRIPTION
**Pull request by NakulSeerpu**
The list of products in api should be either present in db or yaml

### CODE REVIEW

File: src/Kmd.Logic.Gateway.Automation/PreValidation/ApisPreValidation.cs [hunk 1]
```diff
@@ -22,8 +30,24 @@ public IEnumerable<GatewayAutomationResult> ValidateAsync(PublishFileModel publi
                     this.ValidationResults.Add(new GatewayAutomationResult { IsError = true, ResultCode = ResultCode.ValidationFailed, Message = $"Duplicate api names exist" });
                 }
 
+                using var client = this.gatewayClientFactory.CreateClient();
+                var products = client.GetAllProducts(this.options.SubscriptionId, this.options.ProviderId);
+                var nameofproducts = products.Select(x => x.Name).ToList();
+                var combinedproducts = nameofproducts.Union(publishFileModel.Products.Select(y => y.Name).ToList()).ToList();
```
Comment (Araballi) [inline src/Kmd.Logic.Gateway.Automation/PreValidation/ApisPreValidation.cs:36]: remove last .ToList()

File: src/Kmd.Logic.Gateway.Automation/PreValidation/ApisPreValidation.cs [hunk 2]
```diff
@@ -22,8 +30,24 @@ public IEnumerable<GatewayAutomationResult> ValidateAsync(PublishFileModel publi
...
+
                 foreach (var api in publishFileModel.Apis)
                 {
+                    foreach (var version in api.ApiVersions)
+                    {
+                        foreach (var product in version.ProductNames)
```
Comment (Araballi) [inline src/Kmd.Logic.Gateway.Automation/PreValidation/ApisPreValidation.cs:42]: Null check required here

**Review by Araballi — CHANGES_REQUESTED** [review]

File: src/Kmd.Logic.Gateway.Automation/PreValidation/ApisPreValidation.cs (hunk 1 above)
Comment (NakulSeerpu) [reply src/Kmd.Logic.Gateway.Automation/PreValidation/ApisPreValidation.cs]: Removed

File: src/Kmd.Logic.Gateway.Automation/PreValidation/ApisPreValidation.cs (hunk 2 above)
Comment (NakulSeerpu) [reply src/Kmd.Logic.Gateway.Automation/PreValidation/ApisPreValidation.cs]: Checking for null now.

**Review by jaceks2106 — CHANGES_REQUESTED** [review]
Fix warnings

File: src/Kmd.Logic.Gateway.Automation/PreValidation/ApisPreValidation.cs [hunk 3]
```diff
@@ -22,8 +30,32 @@ public IEnumerable<GatewayAutomationResult> ValidateAsync(PublishFileModel publi
...
+                using var client = this.gatewayClientFactory.CreateClient();
+                var products = client.GetAllProducts(this.options.SubscriptionId, this.options.ProviderId);
+                var nameofproducts = products.Select(x => x.Name).ToList();
+                var consolidatedProductNames = nameofproducts.Union(publishFileModel.Products.Select(y => y.Name).ToList());
+
                 foreach (var api in publishFileModel.Apis)
                 {
+                    if (api.ApiVersions == null)
+                    {
+                        continue;
```
Comment (Araballi) [inline src/Kmd.Logic.Gateway.Automation/PreValidation/ApisPreValidation.cs:42]: API should have version error has to come here

Comment (NakulSeerpu) [reply src/Kmd.Logic.Gateway.Automation/PreValidation/ApisPreValidation.cs:43]: Added

**Review by Araballi — APPROVED** [review]

**Review by jaceks2106 — APPROVED** [review]

**Pull request merged**